controllers/pkg/reconcilers/repository/reconciler_test.go

```diff
@@ -196,3 +202,77 @@ func TestDeleteRepo(t *testing.T) {
 		})
 	}
 }
+
+func Test_reconciler_upsertRepo(t *testing.T) {
+	type mockHelper struct {
+		methodName string
+		argType    []string
+		retArgList []interface{}
+	}
+	type fields struct {
+		APIPatchingApplicator resource.APIPatchingApplicator
+		giteaClient           giteaclient.GiteaClient
+		finalizer             *resource.APIFinalizer
+		l                     logr.Logger
+	}
+	type args struct {
+		ctx         context.Context
+		giteaClient giteaclient.GiteaClient
+		cr          *infrav1alpha1.Repository
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		mocks   []mockHelper
+		wantErr bool
+	}{
+		{
+			name:   "User Info reports error",
+			fields: fields{resource.NewAPIPatchingApplicator(nil), nil, nil, log.FromContext(nil)},
+			args:   args{nil, nil, &infrav1alpha1.Repository{}},
+			mocks: []mockHelper{
+				{"GetMyUserInfo", []string{}, []interface{}{nil, nil, fmt.Errorf("error getting User Information")}},
+			},
+			wantErr: true,
+		},
+		{
+			name:   "Repo exists, cr fields blank",
+			fields: fields{resource.NewAPIPatchingApplicator(nil), nil, nil, log.FromContext(nil)},
+			args:   args{nil, nil, &infrav1alpha1.Repository{Status: infrav1alpha1.RepositoryStatus{}}},
+			mocks: []mockHelper{
+				{"GetMyUserInfo", []string{}, []interface{}{&gitea.User{UserName: "gitea"}, nil, nil}},
+				{"GetRepo", []string{"string", "string"}, []interface{}{&gitea.Repository{}, nil, nil}},
+				{"EditRepo", []string{"string", "string", "gitea.EditRepoOption"}, []interface{}{&gitea.Repository{}, nil, nil}},
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &reconciler{
+				APIPatchingApplicator: tt.fields.APIPatchingApplicator,
+				giteaClient:           tt.fields.giteaClient,
+				finalizer:             tt.fields.finalizer,
+				l:                     tt.fields.l,
+			}
+			// The below block being setup and processing of mocks before invoking the function to be tested
```

Comment:
Lines 260 to L271 is code to process the info in mocks struct to setup prior to actual invocation of the function to be tested.